tests/unit/test_layers.py

```diff
@@ -0,0 +1,327 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+import tarfile
+from pathlib import Path
+from typing import List, Tuple
+
+import pytest
+from craft_parts.overlays import overlays
+
+from rockcraft import errors, layers
+
+
+def get_tar_contents(tar_path: Path) -> List[str]:
+    with tarfile.open(tar_path, "r") as tar_file:
+        return tar_file.getnames()
+
+
+def duplicate_dirs_setup(tmp_path) -> Tuple[Path, Path]:
+    """Create a filetree with an upper layer and a fake 'rootfs' structure.
+
+    layer_dir/
+      |- bin/dir1/a.txt
+      |- usr/bin/dir1/b.txt
+    rootfs/
+      |- usr/bin/
+      |- bin ----> usr/bin (symlink)
+
+    returns a tuple with (layer_dir, rootfs).
+    """
+
+    layer_dir = tmp_path / "layer_dir"
+    layer_dir.mkdir()
+
+    # The new layer has "bin/dir1/a.txt" and "usr/bin/dir1/b.txt".
+    (layer_dir / "bin/dir1").mkdir(parents=True)
+    (layer_dir / "bin/dir1/a.txt").touch()
+    (layer_dir / "usr/bin/dir1").mkdir(parents=True)
+    (layer_dir / "usr/bin/dir1/b.txt").touch()
+
+    rootfs_dir = tmp_path / "rootfs"
+    rootfs_dir.mkdir()
+    # In the base layer "bin" is a symlink to "usr/bin"
+    (rootfs_dir / "usr/bin").mkdir(parents=True)
+    (rootfs_dir / "bin").symlink_to("usr/bin")
+
+    return layer_dir, rootfs_dir
+
+
+def test_archive_layer_directories(tmp_path):
```

Comment:
the tests here are all moved over from `test_oci.py`, with no functional changes and only minor fixes: for example, this was originally named `test_add_layer_directories()` and received a fixture to inspect the tarball. Such fixture is not needed anymore because the function receives the tarpath as a parameter.